diff --git a/src/server/game/World/World.cpp b/src/server/game/World/World.cpp
index de605ee..9cf364b 100644
--- a/src/server/game/World/World.cpp
+++ b/src/server/game/World/World.cpp
@@ -1958,6 +1958,25 @@ void World::SetInitialWorldSettings()
         });
     }
 
+    TC_LOG_INFO("server.loading", "Loading Template Talents...");
+    sTemplateNpcMgr->LoadTalentsContainer();
+
+    // Load templates for Template NPC #2
+    TC_LOG_INFO("server.loading", "Loading Template Glyphs...");
+    sTemplateNpcMgr->LoadGlyphsContainer();
+
+    // Load templates for Template NPC #3
+    TC_LOG_INFO("server.loading", "Loading Template Gear for Humans...");
+    sTemplateNpcMgr->LoadHumanGearContainer();
+
+    // Load templates for Template NPC #4
+    TC_LOG_INFO("server.loading", "Loading Template Gear for Alliances...");
+    sTemplateNpcMgr->LoadAllianceGearContainer();
+
+    // Load templates for Template NPC #5
+    TC_LOG_INFO("server.loading", "Loading Template Gear for Hordes...");
+    sTemplateNpcMgr->LoadHordeGearContainer();
+
     uint32 startupDuration = GetMSTimeDiffToNow(startupBegin);
 
     TC_LOG_INFO("server.worldserver", "World initialized in %u minutes %u seconds", (startupDuration / 60000), ((startupDuration % 60000) / 1000));
diff --git a/src/server/scripts/Custom/custom_script_loader.cpp b/src/server/scripts/Custom/custom_script_loader.cpp
index dd4b5e9..5563328 100644
--- a/src/server/scripts/Custom/custom_script_loader.cpp
+++ b/src/server/scripts/Custom/custom_script_loader.cpp
@@ -17,9 +17,11 @@
 
 // This is where scripts' loading functions should be declared:
 
+void AddSC_NPC_Template();
 
 // The name of this function should match:
 // void Add${NameOfDirectory}Scripts()
 void AddCustomScripts()
 {
+    AddSC_NPC_Template();
 }
